### PR TITLE
fix threshold opt - Now opts for min fpr max tpr

### PR DESCRIPTION
**Pull request by HB-Dynamite**
fixing the wild not working threshold opt.
also add visual for new method when verbose == 1 
e.g. with minic data: 
image

### CODE REVIEW

File: igann/igann.py [hunk 1]
```diff
@@ -343,7 +343,9 @@ def fit(self, X, y, val_set=None, eval=None, plot_fixed_features=None, fitted_du
         '''
 
         self._reset_state()
-
+        X_copy = deepcopy(X)
```
Comment (NicoHambauer) [inline igann/igann.py:346]: Memory intense, i suspect double the RAM consumption? May cause problems with large datasets, as we saw some datasets already reach 64GB or 128GB fast then.
Maybe there is another solution to converting the datatype or accessing the same variable?

**Review by NicoHambauer — COMMENTED** [review]
Overall seems good to go. Please have a look at memory consumption again if that is something in your eyes as well that could be fixed.

File: igann/igann.py (hunk 1 above)
Comment (HB-Dynamite) [reply igann/igann.py:346]: Its a point but I do not think its a real concern! Memory consumption of igann is mostly related to the number of regressors  and not the the size of the dataset.  e.g. a 400mb csv with 1million rows and 20float features consumes ~170mb in ram. ref -> ( https://stackoverflow.com/questions/18089667/how-to-estimate-how-much-memory-a-pandas-dataframe-will-need)

Comment (MathiasKraus) [reply igann/igann.py:346]: This only copies the init feature matrix right? I dont think this is too memory intense

File: igann/igann.py [hunk 2]
```diff
@@ -586,33 +590,68 @@ def _print_results(self, counter, counter_no_progress, eval, boost_rate, train_l
...
+        if hasattr(y_train, "detach"):
+            y_train = y_train.detach().cpu().numpy()
+
+        # Compute ROC curve
+        fpr, tpr, thresholds = roc_curve(y_train, y_proba)
+
+        # Create a DataFrame from the FPR, TPR, and thresholds
+        roc_df = pd.DataFrame({"FPR": fpr, "TPR": tpr, "Thresholds": thresholds})
+
+        print(roc_df)
```
Comment (MathiasKraus) [inline igann/igann.py:612]: Can we remove this print command?

File: igann/igann.py [hunk 3]
```diff
@@ -586,33 +590,68 @@ def _print_results(self, counter, counter_no_progress, eval, boost_rate, train_l
...
+        roc_df["Distance"] = np.sqrt((roc_df["FPR"]) ** 2 + (roc_df["TPR"] - 1) ** 2)
+
+        # Find the index of the row with the minimal distance
+        min_distance_index = roc_df["Distance"].idxmin()
+
+        # Get the best threshold at this index
+        best_threshold = roc_df.loc[min_distance_index, "Thresholds"]
+        self.best_threshold = best_threshold
+
+        if self.verbose == 1:
```
Comment (MathiasKraus) [inline igann/igann.py:624]: Can we change this to self.verbose == 2? verbose == 1 was previously used only for textual output

**Pull request closed without merging**